apps/zadmin/tasks.py

```diff
@@ -488,83 +484,125 @@ def fetch_langpacks(path, **kw):
-        if not allowed.match(xpi):
+        if not allowed_file(xpi):
             continue
 
-        lang = os.path.splitext(xpi)[0]
+        fetch_langpack.delay(url, xpi)
 
-        try:
-            req = requests.get(url,
-                               verify=settings.CA_CERT_BUNDLE_PATH)
-
-            if ('content-length' not in req.headers or
-                int(req.headers['content-length']) >
-                    settings.LANGPACK_MAX_SIZE):
-                log.error('[@None] Language pack "%s" too large: %d > %d'
-                          % (xpi, req.headers['content-large'],
-                             settings.LANGPACK_MAX_SIZE))
-                continue
-
-            chunks = []
-            size = 0
-            for chunk in req.iter_content(settings.LANGPACK_MAX_SIZE):
-                size += len(chunk)
-                # `requests` doesn't respect the Content-Length header
-                # so we need to check twice.
-                if size > settings.LANGPACK_MAX_SIZE:
-                    raise Exception('Response to big')
-                chunks.append(chunk)
-        except Exception, e:
-            log.error('[@None] Error fetching "%s" language pack: %s'
-                      % (xpi, e))
-            continue
 
-        upload = FileUpload()
-        upload.add_file(chunks, xpi, size)
+@task(rate_limit='6/s')
+@write
+def fetch_langpack(url, xpi, **kw):
+    try:
+        req = requests.get(url,
+                           verify=settings.CA_CERT_BUNDLE_PATH)
+
+        if ('content-length' not in req.headers or
+            int(req.headers['content-length']) >
+                settings.LANGPACK_MAX_SIZE):
+            log.error('[@None] Language pack "%s" too large: %d > %d'
+                      % (xpi, req.headers['content-large'],
+                         settings.LANGPACK_MAX_SIZE))
+            return
+
+        chunks = []
+        size = 0
+        for chunk in req.iter_content(settings.LANGPACK_MAX_SIZE):
+            size += len(chunk)
+            # `requests` doesn't respect the Content-Length header
+            # so we need to check twice.
+            if size > settings.LANGPACK_MAX_SIZE:
+                raise Exception('Response to big')
+            chunks.append(chunk)
+    except Exception, e:
+        log.error('[@None] Error fetching "%s" language pack: %s'
+                  % (xpi, e))
+        return
+
+    upload = FileUpload()
+    upload.add_file(chunks, xpi, size)
 
-        # Activate the correct locale for the language pack so it
-        # will be used as the add-on's default locale if available.
-        translation.activate(lang)
+    lang = os.path.splitext(xpi)[0]
 
-        guid = 'langpack-%s@firefox.mozilla.org' % lang
+    # Activate the correct locale for the language pack so it
+    # will be used as the add-on's default locale if available.
+    orig_lang = translation.get_language()
+    translation.activate(lang)
+
+    try:
```

Comment:
Could we use https://docs.djangoproject.com/en/1.8/ref/utils/#django.utils.translation.override instead of this gigantic try/finally block?